File: cleanup_unused_photos.py
"""
Unused Employee Photos Cleanup Script

Identifies and removes employee photo files that are not referenced in the database.
Works with MySQL database and Laravel storage structure for employee management.

Usage:
    python cleanup_unused_photos.py [--dry-run] [--force]

Options:
    --dry-run    Show what would be deleted without actually deleting
    --force      Skip confirmation prompt
"""

import os
import sys
import mysql.connector
from pathlib import Path
from datetime import datetime
import argparse
from typing import Set, List, Dict, Any

class DatabaseConfig:
    """Database connection configuration"""
    
    HOST = '127.0.0.1'
    PORT = 3306
    USER = 'root'
    PASSWORD = ''
    DATABASE = 'aegis_ignis'
    
    @classmethod
    def get_config_dict(cls) -> Dict[str, Any]:
        return {
            'host': cls.HOST,
            'port': cls.PORT,
            'user': cls.USER,
            'password': cls.PASSWORD,
            'database': cls.DATABASE
        }

class FileSystemUtils:
    """Utility functions for file system operations"""
    
    @staticmethod
    def format_size(bytes_size: int) -> str:
        """Format bytes to human readable format"""
        for unit in ['B', 'KB', 'MB', 'GB']:
            if bytes_size < 1024.0:
                return f"{bytes_size:.2f} {unit}"
            bytes_size /= 1024.0
        return f"{bytes_size:.2f} TB"


def get_referenced_photos():
    """Get all photo paths referenced in the database"""
    try:
        conn = mysql.connector.connect(**DB_CONFIG)
        cursor = conn.cursor()
        
        query = """
            SELECT face_photo_path, photo_url 
            FROM employees 
            WHERE face_photo_path IS NOT NULL OR photo_url IS NOT NULL
        """
        
        cursor.execute(query)
        results = cursor.fetchall()
        
        used_paths = set()
        for row in results:
            face_photo, photo_url = row
            if face_photo:
                used_paths.add(face_photo)
            if photo_url:
                # Extract path from URL
                path = photo_url.replace('/storage/', '').replace('storage/', '')
                used_paths.add(path)
        
        cursor.close()
        conn.close()
        
        return used_paths
        
    except mysql.connector.Error as err:
        print(f"❌ Database error: {err}")
        sys.exit(1)


def scan_directory(directory, used_paths):
    """Scan a directory for unused files"""
    dir_path = STORAGE_PATH / directory
    
    if not dir_path.exists():
        return []
    
    unused_files = []
    
    for file_path in dir_path.iterdir():
        if not file_path.is_file():
            continue
        
        # Check various path formats
        relative_path = f"{directory}/{file_path.name}"
        
        is_used = (
            relative_path in used_paths or
            f"public/{relative_path}" in used_paths or
            file_path.name in used_paths
        )
        
        if not is_used:
            stat = file_path.stat()
            unused_files.append({
                'path': file_path,
                'relative_path': relative_path,
                'size': stat.st_size,
                'modified': stat.st_mtime
            })
    
    return unused_files


def main():
    parser = argparse.ArgumentParser(
        description='Cleanup unused employee photos from storage'
    )
    parser.add_argument(
        '--dry-run',
        action='store_true',
        help='Show what would be deleted without actually deleting'
    )
    parser.add_argument(
        '--force',
        action='store_true',
        help='Skip confirmation prompt'
    )
    
    args = parser.parse_args()
    
    print("=" * 50)
    print("   UNUSED PHOTOS CLEANUP UTILITY")
    print("=" * 50)
    print()
    
    if args.dry_run:
        print("⚠️  DRY RUN MODE - No files will be deleted\n")
    
    # Get referenced photos from database
    print("📊 Scanning database for referenced photos...")
    used_paths = get_referenced_photos()
    print(f"   Found {len(used_paths)} photos referenced in database\n")
    
    # Scan directories
    all_unused_files = []
    total_files = 0
    
    for directory in DIRECTORIES:
        print(f"📁 Scanning directory: storage/app/public/{directory}")
        
        dir_path = STORAGE_PATH / directory
        if not dir_path.exists():
            print(f"   ⚠️  Directory does not exist, skipping...\n")
            continue
        
        unused_files = scan_directory(directory, used_paths)
        all_unused_files.extend(unused_files)
        
        files_count = len(list(dir_path.glob('*')))
        total_files += files_count
        
        print(f"   Files found: {files_count}")
        print(f"   Unused files: {len(unused_files)}\n")
    
    # Display results
    print("=" * 50)
    print("   SCAN RESULTS")
    print("=" * 50)
    print(f"Total files scanned:  {total_files}")
    print(f"Files in use:         {total_files - len(all_unused_files)}")
    print(f"Unused files:         {len(all_unused_files)}")
    print("-" * 50)
    print()
    
    if not all_unused_files:
        print("✅ No unused files found! Your storage is clean.")
        return
    
    # Display unused files
    print("📋 UNUSED FILES TO BE DELETED:\n")
    total_size = 0
    
    for file_info in all_unused_files:
        date_str = datetime.fromtimestamp(file_info['modified']).strftime('%Y-%m-%d %H:%M:%S')
        size_str = format_size(file_info['size'])
        print(f"   • {file_info['relative_path']}")
        print(f"     Size: {size_str} | Last modified: {date_str}")
        total_size += file_info['size']
    
    print(f"\n💾 Total space to be freed: {format_size(total_size)}\n")
    
    # Delete files
    if not args.dry_run:
        if not args.force:
            print(f"⚠️  WARNING: This will permanently delete {len(all_unused_files)} files!")
            confirmation = input("Do you want to proceed? (yes/no): ")
            
            if confirmation.lower() != 'yes':
                print("\n❌ Operation cancelled.")
                return
        
        print("\n🗑️  Deleting files...\n")
        
        deleted_count = 0
        error_count = 0
        
        for file_info in all_unused_files:
            try:
                file_info['path'].unlink()
                deleted_count += 1
                print(f"   ✓ Deleted: {file_info['relative_path']}")
            except Exception as e:
                error_count += 1
                print(f"   ✗ Error deleting {file_info['relative_path']}: {e}")
        
        print("\n" + "=" * 50)
        print("   CLEANUP COMPLETE")
        print("=" * 50)
        print(f"Files deleted:        {deleted_count}")
        print(f"Errors:               {error_count}")
        print(f"Space freed:          {format_size(total_size)}")
        print("=" * 50)
        
        if error_count == 0:
            print("\n✅ All unused photos have been successfully deleted!")
        else:
            print(f"\n⚠️  Cleanup completed with {error_count} errors.")
    else:
        print("\n✅ Dry run complete. No files were deleted.")
        print("   Run without --dry-run to actually delete files.")


if __name__ == '__main__':
    main()

File: Server/app/Console/Commands/CleanupUnusedPhotos.php
<?php

namespace App\Console\Commands;

use Illuminate\Console\Command;
use Illuminate\Support\Facades\DB;
use Illuminate\Support\Facades\Storage;

class CleanupUnusedPhotos extends Command
{
    
    protected $signature = 'photos:cleanup {--dry-run : Preview without deleting} {--force : Skip confirmation}';

   
    protected $description = 'Automatically cleanup unused employee photos not shown on website';

    
    public function handle()
    {
        $dryRun = $this->option('dry-run');
        $force = $this->option('force');

        $this->info('🧹 Automatic Photo Cleanup - Removing photos not shown on website');
        $this->newLine();

        if ($dryRun) {
            $this->warn('DRY RUN MODE - No files will be deleted');
            $this->newLine();
        }

        // Directories to check
        $directories = ['faces', 'employees', 'detections', 'alerts'];
        
        $stats = [
            'total_files' => 0,
            'used_files' => 0,
            'unused_files' => 0,
            'deleted_files' => 0,
            'errors' => 0,
            'space_freed' => 0
        ];

        // Get all photos referenced in the database (shown on website)
        $this->info('Scanning database for photos shown on website...');
        
        $usedPaths = [];
        
        // From employees table
        $employees = DB::table('employees')
            ->whereNotNull('face_photo_path')
            ->orWhereNotNull('photo_url')
            ->get(['face_photo_path', 'photo_url']);
        
        foreach ($employees as $employee) {
            if ($employee->face_photo_path) {
                $usedPaths[] = $employee->face_photo_path;
            }
            if ($employee->photo_url) {
                $path = str_replace([url('storage/'), '/storage/'], '', $employee->photo_url);
                $usedPaths[] = $path;
            }
        }
        
        // From detection_logs (recent detections shown on website) - if table exists
        try {
            $recentDetections = DB::table('detection_logs')
                ->where('created_at', '>=', now()->subDays(7))
                ->whereNotNull('detection_image_path')
                ->pluck('detection_image_path');
            
            foreach ($recentDetections as $path) {
                $usedPaths[] = $path;
            }
        } catch (\Exception $e) {
            // Table doesn't exist, skip
        }
        
        // From fire_alerts (shown on website) - if table exists
        try {
            $fireAlerts = DB::table('fire_alerts')
                ->where('created_at', '>=', now()->subDays(30))
                ->whereNotNull('image_path')
                ->pluck('image_path');
            
            foreach ($fireAlerts as $path) {
                $usedPaths[] = $path;
            }
        } catch (\Exception $e) {
            // Table doesn't exist, skip
        }
        
        $usedPaths = array_unique($usedPaths);
        $this->info("   Found " . count($usedPaths) . " photos currently shown on website");
        $this->newLine();

        $unusedFiles = [];

        // Scan each directory
        foreach ($directories as $directory) {
            $fullPath = storage_path("app/public/{$directory}");
            
            if (!is_dir($fullPath)) {
                continue;
            }
            
            $this->info("Scanning: storage/app/public/{$directory}");
            
            $files = scandir($fullPath);
            $filesInDir = 0;
            $unusedInDir = 0;
            
            foreach ($files as $file) {
                if ($file === '.' || $file === '..') {
                    continue;
                }
                
                $filePath = "{$fullPath}/{$file}";
                
                if (!is_file($filePath)) {
                    continue;
                }
                
                $filesInDir++;
                $stats['total_files']++;
                
                $relativePath = "{$directory}/{$file}";
                $isUsed = in_array($relativePath, $usedPaths);
                
                if ($isUsed) {
                    $stats['used_files']++;
                } else {
                    $stats['unused_files']++;
                    $unusedInDir++;
                    
                    $fileSize = filesize($filePath);
                    $unusedFiles[] = [
                        'path' => $filePath,
                        'relative_path' => $relativePath,
                        'size' => $fileSize,
                        'modified' => filemtime($filePath)
                    ];
                }
            }
            
            $this->line("   Files: {$filesInDir} | Unused: {$unusedInDir}");
        }
        
        $this->newLine();
        
        // Display results
        $this->info('═══════════════════════════════════');
        $this->info('   SCAN RESULTS');
        $this->info('═══════════════════════════════════');
        $this->line("Total files scanned:  {$stats['total_files']}");
        $this->line("Files shown on site:  {$stats['used_files']}");
        $this->warn("Unused files:         {$stats['unused_files']}");
        $this->newLine();

        if (count($unusedFiles) === 0) {
            $this->info('No unused files found! Storage is clean.');
            return 0;
        }

        // Calculate space
        foreach ($unusedFiles as $file) {
            $stats['space_freed'] += $file['size'];
        }
        
        $spaceMB = round($stats['space_freed'] / 1024 / 1024, 2);
        $this->warn("Space to be freed: {$spaceMB} MB from {$stats['unused_files']} files");
        $this->newLine();

        // Delete files
        if (!$dryRun) {
            if (!$force && !$this->confirm('Delete all unused photos?', true)) {
                $this->error('Operation cancelled.');
                return 1;
            }
            
            $this->info('Deleting unused files...');
            
            $progressBar = $this->output->createProgressBar(count($unusedFiles));
            $progressBar->start();
            
            foreach ($unusedFiles as $file) {
                try {
                    if (unlink($file['path'])) {
                        $stats['deleted_files']++;
                    } else {
                        $stats['errors']++;
                    }
                } catch (\Exception $e) {
                    $stats['errors']++;
                }
                $progressBar->advance();
            }
            
            $progressBar->finish();
            $this->newLine(2);
            
            $this->info('═══════════════════════════════════');
            $this->info('   CLEANUP COMPLETE');
            $this->info('═══════════════════════════════════');
            $this->line("Files deleted:        {$stats['deleted_files']}");
            $this->line("Errors:               {$stats['errors']}");
            $this->line("Space freed:          {$spaceMB} MB");
            $this->info('═══════════════════════════════════');
            $this->newLine();
            
            if ($stats['errors'] === 0) {
                $this->info('All unused photos successfully deleted!');
                
                // Log the cleanup
                \Log::info('Automatic photo cleanup completed', [
                    'deleted' => $stats['deleted_files'],
                    'space_freed_mb' => $spaceMB
                ]);
            } else {
                $this->warn("Completed with {$stats['errors']} errors.");
            }
        } else {
            $this->info('Dry run complete. No files deleted.');
            $this->line('   Run without --dry-run to delete files.');
        }

        return 0;
    }
}
